### PR TITLE
setuptools-based plugin for StatsWriters

### PR DESCRIPTION
**Pull request by chriselion**
### Proposed change(s)
Adds a setuptools-based plugin system, and an interface for registering custom StatsWriters. Also refactored some of the output directory settings so that TensorBoardWriter just needs the RunOptions.

For more background, see [this video](https://www.youtube.com/watch?v=fY3Y_xPKWNA) and [example code](https://github.com/anthonywritescode/explains/tree/master/sample_code/ep128).

Example output:
```console
$ mlagents-learn --force
...
2020-12-21 15:06:51 INFO [learn.py:274] run_seed set to 249
registering default
registering example
Creating a new stats writer! This is so exciting!

...

ExampleStatsWriter category: 3DBall values: {'Policy/Entropy': StatsSummary(mean=1.4183114, std=0.00084836915, num=50172), ...}
```

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)
https://github.com/anthonywritescode/explains/tree/master/sample_code/ep128 
https://www.youtube.com/watch?v=fY3Y_xPKWNA

### Types of change(s)
- [x] New feature

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)

### CODE REVIEW

File: ml-agents-plugin-examples/setup.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from setuptools import setup
```
Comment (chriselion) [inline ml-agents-plugin-examples/setup.py:1]: No plans to publish this package, but we could use it to set up tests, e.g. for bad imports.

File: ml-agents/setup.py [hunk 2]
```diff
@@ -71,13 +71,17 @@ def run(self):
...
+        "importlib_metadata",  # TODO for python<3.8 only
     ],
     python_requires=">=3.6.1",
     entry_points={
         "console_scripts": [
             "mlagents-learn=mlagents.trainers.learn:main",
             "mlagents-run-experiment=mlagents.trainers.run_experiment:main",
-        ]
+        ],
+        "mlagents.stats_writer": [
```
Comment (chriselion) [inline ml-agents/setup.py:82]: We would need to add a new key for each type of plugin interface we want, e.g. `mlagents.demonstration_provider: [...]`

File: ml-agents-plugin-examples/setup.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+setup(
+    name="mlagents_plugin_examples",
+    version="0.0.1",
+    entry_points={
+        "mlagents.stats_writer": [
+            "example=mlagents_plugin_examples.example_stats_writer:get_example_stats_writer"
```
Comment (chriselion) [inline ml-agents-plugin-examples/setup.py:8]: The form of this is `{entry point name}={plugin module}.{plugin_function}`

File: docs/Training-Plugins.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+In the call to `setup()`, you'll need to add to the `entry_points` dictionary for each plugin interface that you
+implement. The form of this is `{entry point name}={plugin module}:{plugin function}`. For example, in
+ `ml-agents-plugin-examples`:
+```python
+"mlagents.stats_writer": [
+    "example=mlagents_plugin_examples.example_stats_writer:get_example_stats_writer"
+]
+```
+* `mlagents.stats_writer` is the name of the plugin interface. This must be one of the provided interfaces (see below).
```
Comment (vincentpierre) [inline docs/Training-Plugins.md:27]: Make the see below a link to line 41

File: docs/Training-Plugins.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+### Local Installation
+Once you've defined `entry_points` in your `setup.py`, you will need to run
+```
+pip install -e [path to your plugin code]
+```
+in the same python virtual environment that you have `mlagents` installed.
+
+## Plugin Interfaces
+
+### StatsWriter
```
Comment (vincentpierre) [inline docs/Training-Plugins.md:43]: It is not obvious that `mlagents.stats_writer` is a magic string. Can this be made more explicit here ?

File: ml-agents/mlagents/plugins/stats_writer.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    ]
+
+
+def register_stats_writer_plugins(run_options: RunOptions) -> List[StatsWriter]:
+    """
+    Registers all StatsWriter plugins (including the default one),
+    and evaluates them, and returns the list of all the StatsWriter implementations.
+    """
+    all_stats_writers: List[StatsWriter] = []
+    entry_points = importlib_metadata.entry_points()["mlagents.stats_writer"]
```
Comment (vincentpierre) [inline ml-agents/mlagents/plugins/stats_writer.py:45]: Can "mlagents.stats_writer" be made a constant / static variable?

File: ml-agents/mlagents/plugins/stats_writer.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+        try:
+            logger.debug(f"Initializing StatsWriter plugins: {entry_point.name}")
+            plugin_func = entry_point.load()
+            plugin_stats_writers = plugin_func(run_options)
+            logger.debug(
+                f"Found {len(plugin_stats_writers)} StatsWriters for plugin {entry_point.name}"
+            )
+            all_stats_writers += plugin_stats_writers
+        except BaseException:
```
Comment (vincentpierre) [inline ml-agents/mlagents/plugins/stats_writer.py:57]: I think catching the error is great, but the error should be displayed as well / be logged somewhere.

File: ml-agents/mlagents/trainers/cli_utils.py [hunk 8]
```diff
@@ -189,6 +189,9 @@ def _create_parser() -> argparse.ArgumentParser:
         action=RaiseRemovedWarning,
         help="(Removed) Use the TensorFlow framework.",
     )
+    argparser.add_argument(
+        "--results-dir", default="results", help="Results base directory"
+    )
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/cli_utils.py:194]: Is this relevant to this PR? Why are we adding this flag ?

**Review by vincentpierre — APPROVED** [review]

File: docs/Training-Plugins.md (hunk 5 above)
Comment (chriselion) [reply docs/Training-Plugins.md:43]: I moved mlagents.stats_writer to a constant in `mlagents.plugins.__init__.py`

File: ml-agents/mlagents/plugins/stats_writer.py (hunk 7 above)
Comment (chriselion) [reply ml-agents/mlagents/plugins/stats_writer.py:57]: `logging.exception` will print the exception stack trace and information:
``` python
>>> import logging
>>> logger = logging.getLogger(__name__)
>>> try:
...   int('foo')
... except:
...   logger.exception("oops")
... 
oops
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
ValueError: invalid literal for int() with base 10: 'foo'
```

File: ml-agents/mlagents/trainers/cli_utils.py (hunk 8 above)
Comment (chriselion) [reply ml-agents/mlagents/trainers/cli_utils.py:194]: This was previously hardcoded in learn.py
https://github.com/Unity-Technologies/ml-agents/pull/4788/files#diff-4e0d7b8aef419254e6fc3c63f5baa53f222afda7c908705e857a22f7d4792c47L68

and is now used in CheckpointSettings.maybe_init_path()
https://github.com/Unity-Technologies/ml-agents/pull/4788/files#diff-546e90789e914f8707fd97391f78c4bba39ae69a965858bbb69c2d324db1ec51R719

I can keep it hardcoded if you prefer, but I do think moving all the path logic to part of RunOptions is a win (and basically a pre-req to make TensorBoardWriter act like a plugin)

File: docs/Training-Plugins.md (hunk 4 above)
Comment (chriselion) [reply docs/Training-Plugins.md:27]: Done

File: ml-agents/mlagents/plugins/stats_writer.py (hunk 6 above)
Comment (chriselion) [reply ml-agents/mlagents/plugins/stats_writer.py:45]: Done

File: ml-agents/mlagents/trainers/cli_utils.py (hunk 8 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/cli_utils.py:194]: If a user changes this from the default `results` will it mess with how they're saved in cloud? cc: @hvpeteet

Comment (hvpeteet) [reply ml-agents/mlagents/trainers/cli_utils.py:194]: If a user can set it in the run_options then yes this could mess with cloud since we don't already take it into account. I don't think it is a reason not to make the change though. We can add logic to overwrite this field since we already modify run_options. 

Just a couple of questions to confirm my understanding:

1. Would this be set under `checkpoint_settings --> write_path`?
2. Could we blindly set this for every experiment and it would be ignored by older mlagents versions?

Comment (chriselion) [reply ml-agents/mlagents/trainers/cli_utils.py:194]: 1. It would be `checkpoint_settings -> results_dir` (at the top level)
2. This would trigger an error here https://github.com/Unity-Technologies/ml-agents/blob/5db6efebad7c27ea0437a37f038214a09d9e9d26/ml-agents/mlagents/trainers/settings.py#L37

Comment (hvpeteet) [reply ml-agents/mlagents/trainers/cli_utils.py:194]: Thanks, we can check for it and only overwrite it if already set then. When you check this in can you file a ticket against me to make sure we get this added at some point?

Comment (chriselion) [reply ml-agents/mlagents/trainers/cli_utils.py:194]: Will do
